Add new Core Concepts and Workflow Core docs content.

zach-klippenstein opened the pull request:
First chunk of work for #478.

davidapgar commented on docs/userguide/core-workflow.md:
```diff
@@ -97,73 +108,180 @@ struct State {
+        title: "A nice title",
+        onTap: { sink.send(Action.refreshButtonTapped) }
+}
+TK
```
`TK` likely not needed anymore

zach-klippenstein replied:
fixed.

bencochran commented on docs/userguide/core-patterns.md:
```diff
@@ -1,3 +1,70 @@
+
+### Swift
+
+A Swift workerflow typically receives its dependencies as initializer arguments, just like its input
```
`workerflow` -> `workflow`

zach-klippenstein replied:
**workerflow** – a workflow that just wraps a worker? 😅

bencochran commented on docs/userguide/core-worker.md:
```diff
@@ -1,3 +1,61 @@
+
+While Workers are _backed_ by reactive streams with library-specific subscription APIs, you never
+actually subscribe directly to a worker yourself. Instead, a Workflow asks the infrastructure to
+listen to a worker, and the infrastructure will take care of initializing and tearing down the
```
maybe `run` instead of `listen to`?

bencochran commented on docs/userguide/core-worker.md:
```diff
@@ -1,3 +1,61 @@
+A workflow can ask the infrastructure to await the result of a worker by handing that worker to the
+`RenderContext` within a call to the `render` method. A workflow can handle outputs from a `Worker`.
+
+## Workers provide a declarative window into the imperative world
```
I’m not sure exactly where, but it may be nice to fit something about “workers are a way to perform side effects” into this paragraph somewhere. Mostly trying to give people who have some preexisting mental models something to attach them too.

zach-klippenstein replied:
> ## Workers can perform side effects
> 
> Unlike workflows' `render` method, which can be called many times and must be idempotent, workers
> are started and then ran until completion (or cancellation) – independently of how many times the
> workflow running them is actually rendered. This means that side effects that should be performed
> only once when a workflow enters a particular state, for example, should be placed into a `Worker`
> that the workflow runs while in that state.

bencochran commented on docs/userguide/core-workflow.md:
```diff
@@ -97,73 +108,180 @@ struct State {
+function as a map of `Input` + `State` + childrens' `Rendering`s -> `Rendering`. While a workflow's
+internal state may contain more detailed or comprehensive state, the `Rendering` (external state) is
+a type that is useful outside of the workflow.
 
```
Consider putting something like:

> Because a workflow’s `render` method may be called by infrastructure for a variety of reasons, it’s important to not perform side effects when rendering. Event-based side effects should use Actions and state-based side effects should use Workers.

rjrjr replied:
> …important to not perform side effects when rendering — render methods _must_ be idempotent.

bencochran commented on docs/userguide/core-workflow.md:
```diff
@@ -97,73 +108,180 @@ struct State {
-If a workflow is producing a view model, it is common to need an event handler to respond to UI
-events. The `RenderContext` has API to create an event handler that, when called, will advance the
-workflow by dispatching an action back to the workflow.
+    It is an error to render workflows of the same type more than once in the same render pass.
```
Since we give you a way to render the same type, it feels strange to say it’s an error.

Maybe something like “It is an error to render the same child workflow more than once in the same. Since type is used for workflow identity, the child rendering APIs take an optional string key to differentiate between multiple child workflows of the same type.”

rjrjr commented:
Did I mention yet that this is so good? This is so good,

zach-klippenstein commented on docs/userguide/concepts.md:
```diff
@@ -1,64 +1,108 @@
+
+Web UI is traditionally [declarative](https://en.wikipedia.org/wiki/Declarative_programming) – it
+describes what to render, and some aspects of how to render it (style), but doesn’t say how to
+actually draw it. Declarative code usually easier to read than imperative code the code directly
```
Fixed (both)

zach-klippenstein commented on docs/userguide/concepts.md:
```diff
@@ -1,64 +1,108 @@
+    Kotlin" blurbs. See [Where Swift and Kotlin Libraries Differ](4_where_swift_and_kotlin_libraries_differ.md)
+    for an overall summary.
+
+    In general, any time a generic type is referred to as `Foo`, in source code the Swift associated
```
Changed to 

> In general, any time a generic type is referred to as `Foo`, in source code the Swift associated type is called `Foo` and the Kotlin type parameter is called `FooT`.

zach-klippenstein commented on docs/userguide/core-patterns.md:
```diff
@@ -1,3 +1,70 @@
+## Stateless Workflows
+
+Remember that workflow state is made up of public and private parts. When a workflow's state
+consists entirely of public state, it can ignore all the machinery for private state. In Swift, the
```
fixed

zach-klippenstein commented on docs/userguide/core-patterns.md:
```diff
@@ -1,3 +1,70 @@
+`State` type can be `Void`, and in `Kotlin` it can be `Unit` – such workflows are often referred to
+as "stateless", since they have no state of their own.
+
+## Inputless Workflows
```
fixed

zach-klippenstein commented on docs/userguide/core-patterns.md:
```diff
@@ -1,3 +1,70 @@
+
+## Inputless Workflows
+
+Some workflows manage all of their state internally, and have no public state (aka input). In Swift,
```
fixed

zach-klippenstein commented on docs/userguide/core-worker.md:
```diff
@@ -1,3 +1,61 @@
-_Coming soon!_
+## The Role of a Worker
+
+`Worker` is a protocol (in Swift) and interface (in Kotlin) that defines an asynchronous task that
```
why?

zach-klippenstein commented on docs/userguide/core-worker.md:
```diff
@@ -1,3 +1,61 @@
+They are similar to child workflows with `Void`/`Unit` rendering types.
+
+A workflow can ask the infrastructure to await the result of a worker by handing that worker to the
+`RenderContext` within a call to the `render` method. A workflow can handle outputs from a `Worker`.
```
> A workflow can ask the infrastructure to await the result of a worker by passing that worker to the
`RenderContext.runningWorker` method within a call to the `render` method. A workflow can handle
outputs from a `Worker`.

zach-klippenstein commented on docs/userguide/core-worker.md:
```diff
@@ -1,3 +1,61 @@
+allow wrapping imperative APIs so that Workflows can interact with them in a declarative fashion.
+Instead of making imperative "start this, do that, now stop" calls, a Workflow can say "I declare
+that this task should now be running" and let the infrastructure worry about ensuring the task is
+actually started when necessary, and torn down when it's not needed anymore.
```
Fixed.

zach-klippenstein commented:
Thanks for the thorough reviews yall! @bencochran @davidapgar @rjrjr I addressed all your comments, will merge when this PR goes green. Feel free to re-review and we can address in a follow up if I've merged already.
